API improvements around Executor/Executable

jglick opened the pull request:
See https://github.com/jenkinsci/workflow-plugin/pull/95 for main use cases.
- [X] [JENKINS-26090](https://issues.jenkins-ci.org/browse/JENKINS-26090): `Executables.getExecutor`
- ~~[JENKINS-26092](https://issues.jenkins-ci.org/browse/JENKINS-26092): `executors.jelly` interface underspecified~~
- [x] [JENKINS-25938](https://issues.jenkins-ci.org/browse/JENKINS-25938): lock an `Executor` without creating a `Thread`
- [x] [JENKINS-22941](https://issues.jenkins-ci.org/browse/JENKINS-22941): way to mark an `Executable` that should not block `isReadyToRestart`
- [x] [JENKINS-26900](https://issues.jenkins-ci.org/browse/JENKINS-26900): hide flyweight master executor when ≥1 heavyweight executors running as subtasks

@reviewbybees

jglick committed:
[FIXED JENKINS-26090] Executables.getExecutor

jglick committed:
Use ResourceList.EMPTY.

jglick committed:
[JENKINS-25938] Introduced AsynchronousExecutable.

jglick committed:
Merge branch 'master' into Executable-JENKINS-25938

jglick committed:
Added interruptForShutdown flag.
Long-running executions can ignore shutdown events.

jglick committed:
[JENKINS-26092] Noting some things that could be improved.

jglick committed:
Abandoning Executable2; doing everything in AsynchronousExecution.

jglick committed:
[FIXED JENKINS-22941] AsynchronousExecution.blocksRestart

jglick committed:
[FIXED JENKINS-26900] AsynchronousExecution.displayCell

jglick committed:
Deleting forgotten reference to Executable2.

jtnord commented:
a :+1: from syntax etc - but unfamiliar with the code.

jtnord commented:
is throwing exceptions to indicate normal behavior the only way?
If it has come to this by keeping Jenkins API compatible is it not time to stand back and think about doing something different rather than bending things so that new and previously unthought of constructs can be shoe horned into the core?

jglick commented:
> is throwing exceptions to indicate normal behavior the only way?

Admittedly not ideal but seemed the simplest way to deal with an existing interface that defined a synchronous `void run()` method. In fact if you look at the new implementations, they start off running synchronously in the executor thread, and then _if all goes well_ they switch to asynchronous mode for the remainder of the work. The executables can still take advantage of existing infrastructure surrounding `Executor`, they just need to avoid consuming a native thread, and make some minor customizations to other behaviors.

stephenc commented on core/src/main/java/hudson/model/ResourceController.java:
```diff
@@ -88,6 +88,7 @@ public void execute(@Nonnull Runnable task, ResourceActivity activity ) throws I
         try {
             task.run();
         } finally {
+            // TODO if AsynchronousExecution, do that later
```
You may need to revisit the locking after #1596

stephenc commented on core/src/main/resources/lib/hudson/executors.jelly:
```diff
@@ -39,8 +39,9 @@ THE SOFTWARE.
       <j:if test="${!c.acceptingTasks}"> <st:nbsp/> (${%suspended})</j:if>
     </d:tag>
     <d:tag name="executor">
+      <j:set var="ae" value="${e.asynchronousExecution}"/>
```
revisit against race conditions after #1596 is merged

jglick replied:
I doubt that should have any relationship. If `asynchronousExecution` is set, the executor has already started, so any queue changes are long done. This is only displaying a snapshot of the executor state anyway.

stephenc replied:
And that was where I caught my NPEs... (the issue being that `isIdle()` can have changed *after* you have captured `ae`... really need an atomic state snapshot object that executor returns in order to display its state consistently

jglick replied:
If we capture `ae` and then it goes idle, there is no harm done that I can see. But yes it would be nice to incorporate into `DisplayExecutor`.

Peggs replied:
How do I stop getting these emails?

Sent from my iPhone

> On 25 Mar 2015, at 15:25, Jesse Glick <notifications@github.com> wrote:
> 
> In core/src/main/resources/lib/hudson/executors.jelly:
> 
> > @@ -39,8 +39,9 @@ THE SOFTWARE.
> >        <j:if test="${!c.acceptingTasks}"> <st:nbsp/> (${%suspended})</j:if>
> >      </d:tag>
> >      <d:tag name="executor">
> > +      <j:set var="ae" value="${e.asynchronousExecution}"/>
> If we capture ae and then it goes idle, there is no harm done that I can see. But yes it would be nice to incorporate into DisplayExecutor.
> 
> —
> Reply to this email directly or view it on GitHub.
>

orrc replied:
@Peggs At the top of any page in this project, click the "Unwatch" button and choose "Not watching".

stephenc commented:
:+1:

jtnord commented:
I am ok (+1) on the code.
I am -.5 on the design of throwing exceptions to indicate asynchronous work - but can be persuaded to a +.5 if it is the only way it can be done in a backwards compatible way.

jglick commented:
> if it is the only way it can be done in a backwards compatible way

It is certainly not the only way; it felt like the most straightforward way I could come up with. Providing

```java
interface Executable2 extends Executable {
    AsynchronousExecution run2();
}
```

would work, but then implementations still need to implement `void run()` to do something—`assert false;`? Making `Executable2` not be related to `Executable` is not reasonable, because then there is no well-defined type for executables, which various parts of core code pass around.

Also for the known implementations, it is actually convenient to use the `Executor` thread which is already there to do the preparatory work (that sometimes actually fails with a regular exception or returns early), which would otherwise need to farmed out to an executor service or the like.

Ideally the original design would have not made `Executor extends Thread` to begin with, so `Executor` would merely be a logical slot, and you could decide to use whatever thread  pool you liked to run your body. This would be a big and probably incompatible refactoring.

(Saving the allocation of a short-lived native thread would also be a benefit, but according to my tests on Ubuntu/Java 8 a trivial one: I compared a program that constructed 1_000_000 `Thread`s and did not use them, to one which `start`ed them all but let them finish after a 50ms `sleep`, to one which `start`ed them and forced them to stay alive. The last of course got a process limit error around 30k threads. The first was instantaneous: `Thread.<init>` is basically free. I expected the second to be a lot slower, but in fact it only took a couple minutes including logging—negligible compared to other overheads involved in creating and running builds. `jvisualvm` confirmed that at any given time there were only a few dozen threads remaining in the “dead” state; `jstack` does not show them.)

kohsuke commented:
One of the earlier prototypes of servlet 3.0 (and IIUC Jetty did it this way) was to throw an exception to indicate asynchronous processing, so I think there's some known precedent.

Another alternative is to do what servlet 3.0 is doing in the final version, that you set a flag before returning to indicate that you intend the processing to be asynchronous.

I think the current approach (of indicating the intent by throwing an exception) is fine.

kohsuke commented on core/src/main/java/hudson/model/queue/Executables.java:
```diff
@@ -71,5 +73,33 @@ public static long getEstimatedDurationFor(Executable e) {
+     * @return the executor (possibly a {@link OneOffExecutor}) whose {@link Executor#getCurrentExecutable} matches that, or null
+     * @since TODO
+     */
+    public static @CheckForNull Executor getExecutor(Executable executable) {
```
Feels like this belongs more to the `Executor` class, like `Executor.of(executable)`

kohsuke committed:
javadoc formatting fix for better readability from code editor

kohsuke committed:
javadoc improvement

jglick committed:
Merge branch 'master' into Executable-JENKINS-25938

kohsuke committed:
More minor javadoc

jglick committed:
As requested by @kohsuke, rename Executables.getExecutor to Executor.of.

kohsuke committed:
Fixed a race condition.

IIUC, the expectation here is that Queue.Executable will instantiate AsynchronousException as a handle, start something asynchronously (say send a JMS message) with a callback, and the callback will tickle AsynchronousException.

So AsynchronousException might complete before it gets its executor set. This code change ensures that that case gets handled correctly.

jglick committed:
Merge pull request #1 from jenkinsci/pull-1610

Additional change suggestions on top of PR 1610

kohsuke commented:
:+1:

stephenc commented:
I think this should be merged after #1596 given there are some race condition changes from @kohsuke

jglick commented:
> given there are some race condition changes from @kohsuke

63d3b26 does not have anything to do with `Queue` locking; only applies once the build is underway.

stephenc commented on core/src/main/java/jenkins/model/queue/AsynchronousExecution.java:
```diff
@@ -0,0 +1,122 @@
+ */
+public abstract class AsynchronousExecution extends RuntimeException {
+
+    private Executor executor;
```
`@GuardedBy("this")` if it really is guarded... ditto for `result`

stephenc commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3492,16 +3491,13 @@ public void run() {
                     }
-                } catch (InterruptedException e) {
+                } catch (Exception e) {
                     LOGGER.log(Level.WARNING, "Failed to shutdown Hudson",e);
```
When touching the line above I would have done an `s/Hudson/Jenkins/`

stephenc commented on core/src/main/java/hudson/model/Executor.java:
```diff
@@ -594,6 +653,32 @@ protected Object call(Object o, Method m, Object[] args) throws Throwable {
     }
 
     /**
+     * Finds the executor running a given process.
```
Finds the executor that was running a given process while this method was being called. There is the potential that the process may complete by the time this method returns and thus a non-null result could correspond to an Executor running a different process

jglick replied:
I do not think so. As of 1.536, an `Executor` is thrown out after each use. Thus at worst it will return null for something that _was_ running earlier. But this deserves clarification.

stephenc replied:
Well you can still have `Executor executable = Executor.of(executable); assert executable.getCurrentExecutable() == executable` give an Assertion error` as the executable can finish concurrent with returning from the method.

stephenc commented on core/src/main/java/hudson/model/Executor.java:
```diff
@@ -594,6 +653,32 @@ protected Object call(Object o, Method m, Object[] args) throws Throwable {
+     * @return the executor (possibly a {@link OneOffExecutor}) whose {@link Executor#getCurrentExecutable} matches that, or null
+     * @since TODO
+     */
+    public static Executor of(Executable executable) {
```
`@CheckForNull`

jglick committed:
@GuardedBy as suggested by @stephenc, and some internal Javadoc.

jglick committed:
@CheckForNull as suggested by @stephenc.

kohsuke commented:
:+1:

jglick committed:
Addressing some review comments from @stephenc.

recampbell commented:
:+1:

stephenc commented:
:+1:
